### PR TITLE
[intel-oneapi-mkl] provide omp lib for threads=openmp

### PR DESCRIPTION
**Pull request by rscohn2**
Provide omp library when using `intel-oneapi-mkl threads=openmp`

Code is based on implementation from intel.py. The openmp depends on the compiler so make the compiler for the intel-oneapi-mkl install match the openmp you want to use for the application.

Hoping to resolve the issues in #42310 #42368

### CODE REVIEW

**Review by msimberg — COMMENTED** [review]
Thanks @rscohn2! With @albestro and @RMeli we were discussing doing something in this direction to get things working sooner rather than waiting for a perfect solution, but you beat us to it.

A couple of questions:
- Any particular reason to deviate from the `intel.py` version of `openmp_libs`? And if this version has some benefits, perhaps it makes sense to use this in `intel.py` as well.
- Is there official documentation on clang not being supported? I know clang is problematic with later versions not reporting the location of its `libomp` but I'm wondering if the omission is due to that or intel officially not supporting it?
- Do you think it'd make sense to represent the supported compilers with `requires`/`conflicts` instead?

**Comment by msimberg** [comment]
@rscohn2 just for reference, here is the branch that we were considering opening a PR for: https://github.com/spack/spack/compare/develop...msimberg:spack:intel-packages-clang-libomp-path. The last commit is my attempt at a fallback in case `-print-file-name` doesn't print anything useful for clang. It seems to work just fine at least for an ubuntu system. I much prefer your solution of erroring out if nothing can be found instead of passing on the useless `libomp.so`, but I think the fallback can be a useful addition. I'm happy to add that in a separate PR and not at all a blocker for this PR.

**Comment by rscohn2** [comment]
> * Any particular reason to deviate from the `intel.py` version of `openmp_libs`? And if this version has some benefits, perhaps it makes sense to use this in `intel.py` as well.

The main difference is that intel.py uses the libiomp5 that comes with mkl, and oneapi.py uses the libiomp5 that comes with the compiler. I expect either one should work in most cases. But it is possible that they would be different and that would cause unpredictable behavior depending on whether multiple packages were using omp and which one was first in the link line. I am reluctant to change intel.py because it is used for the parallel studio mkl, which has been deprecated.

> * Is there official documentation on clang not being supported? I know clang is problematic with later versions not reporting the location of its `libomp` but I'm wondering if the omission is due to that or intel officially not supporting it?

I asked a MKL developer if clang omp is supported and will let you know. Since it seemed to be supported in intel.py, I did try to get it to work. I had a lot of trouble finding the right combination of packages to install on ubuntu to make the --print-file-path work for clang. I also spoke to the intel compiler maintainer for icpx about the problem with --print-file-path and clang not working and he said it needs to be able to find the library to report the path. This is odd because linking with -lomp worked but --print-file-path did not. It seems like there is a bug in clang driver or something special about the way the library is handled.

> * Do you think it'd make sense to represent the supported compilers with `requires`/`conflicts` instead?

Can you show me what you mean?

**Comment by rscohn2** [comment]
> @rscohn2 just for reference, here is the branch that we were considering opening a PR for: [develop...msimberg:spack:intel-packages-clang-libomp-path](https://github.com/spack/spack/compare/develop...msimberg:spack:intel-packages-clang-libomp-path). The last commit is my attempt at a fallback in case `-print-file-name` doesn't print anything useful for clang. It seems to work just fine at least for an ubuntu system. I much prefer your solution of erroring out if nothing can be found instead of passing on the useless `libomp.so`, but I think the fallback can be a useful addition. I'm happy to add that in a separate PR and not at all a blocker for this PR.

If people need libomp, then I would appreciate help getting it to work. Do you use it or know others who do?

**Comment by msimberg** [comment]
> > * Any particular reason to deviate from the `intel.py` version of `openmp_libs`? And if this version has some benefits, perhaps it makes sense to use this in `intel.py` as well.
> 
> The main difference is that intel.py uses the libiomp5 that comes with mkl, and oneapi.py uses the libiomp5 that comes with the compiler. I expect either one should work in most cases. But it is possible that they would be different and that would cause unpredictable behavior depending on whether multiple packages were using omp and which one was first in the link line. I am reluctant to change intel.py because it is used for the parallel studio mkl, which has been deprecated.

Sounds reasonable to me. In that case no need to make `intel.py` and `oneapi.py` the same.

> > * Is there official documentation on clang not being supported? I know clang is problematic with later versions not reporting the location of its `libomp` but I'm wondering if the omission is due to that or intel officially not supporting it?
> 
> I asked a MKL developer if clang omp is supported and will let you know. Since it seemed to be supported in intel.py, I did try to get it to work. I had a lot of trouble finding the right combination of packages to install on ubuntu to make the --print-file-path work for clang. I also spoke to the intel compiler maintainer for icpx about the problem with --print-file-path and clang not working and he said it needs to be able to find the library to report the path. This is odd because linking with -lomp worked but --print-file-path did not. It seems like there is a bug in clang driver or something special about the way the library is handled.

Oh yes, we've fought with this one as well. The workaround in that branch that I linked is to check if `libomp.so` happens to be in a reasonable location, and if it is there use that. Note that this would only be used if `-print-file-name` fails, so it's really a fallback.

> > * Do you think it'd make sense to represent the supported compilers with `requires`/`conflicts` instead?
> 
> Can you show me what you mean?

You should be able to add a
```
    requires("%gcc", "%clang", "%intel", "%oneapi", policy="one_of", msg="reason")
```
somewhere in the `IntelOneApiPackage` definition (close to the variants and conflicts seems idiomatic enough). I just tested it and it seems to work. Note though that you need to import `requires` together with `conflicts` and `variant`: https://github.com/spack/spack/blob/f0146acf005866333812386987f26097a1407841/lib/spack/spack/build_systems/oneapi.py#L17. Edit: the benefit of this is that this will fail during concretization, rather than during the build.

> If people need libomp, then I would appreciate help getting it to work. Do you use it or know others who do?

Not in production but we use it for testing. We haven't used it more partly because of this exact issue though.

**Comment by rscohn2** [comment]
> I'm happy to add that in a separate PR and not at all a blocker for this PR.

The MKL developer says that libomp is not supported, but may work. It sounds like the issue is testing and not a problem with the way it works.  A followup PR to add your fallback would be good. Can you look over this PR again and approve if you think it is ok?

File: lib/spack/spack/build_systems/oneapi.py
Comment (msimberg) [inline lib/spack/spack/build_systems/oneapi.py:222]: Nit, for consistency? Not a blocker.
```suggestion
            raise_lib_error("OneApi package with OpenMP requires one of %gcc, %oneapi,or %intel")

        # query the compiler for the library path
        with self.compiler.compiler_environment():
            omp_lib_path = Executable(self.compiler.cc)(
                "--print-file-name", f"{libname}.{dso_suffix}", output=str
            ).strip()
        # if the compiler cannot find the file, it returns the input path
        if not os.path.exists(omp_lib_path):
            raise_lib_error(f"Cannot locate OpenMP library: {omp_lib_path}")

        omp_libs = LibraryList(omp_lib_path)
        tty.info(f"OneApi requires OpenMP library: {omp_libs}")
```

File: lib/spack/spack/build_systems/oneapi.py
Comment (msimberg) [inline lib/spack/spack/build_systems/oneapi.py:29]: As far as I can tell you only ever pass one string to this function so consider using `raise InstallError` directly?

**Review by msimberg — APPROVED** [review]
Just a couple of style comments, neither blocking.

I haven't tested this myself, but it looks reasonable to me.

I'll follow up with separate PRs to add the `requires` and clang support.

File: lib/spack/spack/build_systems/oneapi.py
Comment (rscohn2) [reply lib/spack/spack/build_systems/oneapi.py:29]: I had copied it over from intel.py because it was used by the openmp code. You are right, it is overkill for its use in oneapi.py. I agree with the other comments. I expect to add openmp support to intel-oneapi-dal in the future. I set automerge so it already merged with the approval. I can make the changes after your PR is done.

**Pull request merged**
